helix-core/src/date.rs

```diff
@@ -0,0 +1,422 @@
+                date: Date::new(2021, 11, 15).unwrap(),
+                range: Range::new(0, 10),
+                field: DateField::Year,
+                format: FORMATS[0],
+            })
+        );
+    }
+
+    #[test]
+    fn test_date_empty_range_at_in_middle() {
+        let rope = Rope::from_str("2021-11-15");
+        let range = Range::point(5);
+        assert_eq!(
+            DateIncrementor::from_range(rope.slice(..), range),
+            Some(DateIncrementor {
+                date: Date::new(2021, 11, 15).unwrap(),
+                range: Range::new(0, 10),
+                field: DateField::Month,
+                format: FORMATS[0],
+            })
+        );
+    }
+
+    #[test]
+    fn test_date_empty_range_at_end() {
+        let rope = Rope::from_str("2021-11-15");
+        let range = Range::point(10);
+        assert_eq!(DateIncrementor::from_range(rope.slice(..), range), None);
+    }
+
+    #[test]
+    fn test_invalid_dates() {
+        let tests = [
+            "0000-00-00",
+            "1980-2-21",
+            "1980-12-1",
+            "12345",
+            "2020-02-30",
+            "1999-12-32",
+            "19-12-32",
+            "1-2-3",
+            "0000/00/00",
+            "1980/2/21",
+            "1980/12/1",
+            "12345",
+            "2020/02/30",
+            "1999/12/32",
+            "19/12/32",
+            "1/2/3",
+        ];
+
+        for invalid in tests {
+            let rope = Rope::from_str(invalid);
+            let range = Range::new(0, 1);
+
+            assert_eq!(DateIncrementor::from_range(rope.slice(..), range), None);
+        }
+    }
+
+    #[test]
+    fn test_increment_dates() {
+        let tests = [
+            // (original, cursor, amount, expected)
+            ("2020-02-28", 0, 1, "2021-02-28"),
+            ("2020-02-29", 0, 1, "2021-03-01"),
+            ("2020-01-31", 5, 1, "2020-02-29"),
+            ("2020-01-20", 5, 1, "2020-02-20"),
+            ("2020-02-28", 8, 1, "2020-02-29"),
+            ("2021-02-28", 8, 1, "2021-03-01"),
+            ("2021-02-28", 0, -1, "2020-02-28"),
+            ("2021-03-01", 0, -1, "2020-03-01"),
+            ("2020-02-29", 5, -1, "2020-01-29"),
+            ("2020-02-20", 5, -1, "2020-01-20"),
+            ("2020-02-29", 8, -1, "2020-02-28"),
+            ("2021-03-01", 8, -1, "2021-02-28"),
+            ("1980/12/21", 8, 100, "1981/03/31"),
+            ("1980/12/21", 8, -100, "1980/09/12"),
+            ("1980/12/21", 8, 1000, "1983/09/17"),
+            ("1980/12/21", 8, -1000, "1978/03/27"),
+        ];
+
+        for (original, cursor, amount, expected) in tests {
+            let rope = Rope::from_str(original);
+            let range = Range::new(cursor, cursor + 1);
+            assert_eq!(
+                DateIncrementor::from_range(rope.slice(..), range)
+                    .unwrap()
+                    .incremented_text(amount),
+                expected.into()
+            );
+        }
+    }
+}
```

Comment:
Can you try to add a test with two dates sticking to each other even if it breaks? From what I see, the current regular expression logic may be able to cover two overlapping dates, it can fail but it's fine for now.

Reply:
Do you mean something like "2020-02-282021-02-28", or something else?

Reply:
Yeah.

Reply:
OK

Reply:
How would you like this to be handled? Currently for "2020-02-282021-02-28", both parts ("2020-02-28" and "2021-02-28") are treated as dates and can be incremented. Is this what we want, or would it make more sense to not increment at all for this case?